Improve Theme Structure

Rohitth007 opened the pull request:
This PR transitions to a different structure for specifying themes. A new themes folder is created that contains all the themes and instructions on how to create a new one.
The commits are as follows:
* **Commit1**: themes: Update Gruvbox color codes.
* **Commit2**: script: To help with creating new structure. TO BE DELETED
* **Commit3**: theme: Add theme file template.
* **Commit4**: theme: Create new zt_dark theme structure.
* **Commit5**: theme: Create new zt_light theme structure.
* **Commit6**: theme: Create new zt_blue theme structure.
* **Commit7**: theme: Create new zt_gruvbox theme structure.
* **Commit8**: theme: Parse theme file to generate urwid theme. 
* **Commit9**: theme: helper: Use a helper function to easily add theme properties.
* **Commit10**: type check enum - WIP(MyPy ERRORS) - To be squashed into Commit 9
* **Commit11**: tests: Test new themes with old themes. TO BE DELETED
* **Commit12**: theme: Delete old themes.
* **Commit13**: theme: Add missing styles to REQUIRED_STYLES and check completeness.

Rohitth007 commented:
I've run into a circular import issue in `themes.py` and `gruvbox.py` I've fixed it by moving import to the bottom but is there a better solution?

neiljp reviewed (commented):
@Rohitth007 Extracting the colors - however we end up storing them - makes this a lot more compact and readable :+1:
I left some ideas on adjusting the structure further which could simplify/clarify the processing.

It would be useful to see some better commit structure here - this looks just like the way you're layering on the changes? I just reviewed the net result ("files"); intermediate states of development can be useful, but generally reviewers are more interested in how you're layering the changes upon each other intentionally rather than over time (at least in a branch/PR).

While the automation using the script is good to see, if we're going to migrate to this then I'll reiterate that I would like to see a programmatic test that compares the existing and new results for themes at each color depth - rather than relying upon manual comparisons. We might only have that test present after the first few commit(s) and then remove it with the old code, but it would ensure that the new code is good while we have both present.

Layout-wise, I'm not sure yet where we'll want the end result, but for now you could consider putting the generate-theme code in `run.py` to avoid the import issue?

rht commented on zulipterminal/config/gruvbox.py:
```diff
@@ -0,0 +1,73 @@
+
+
+# fmt: off
+class Color(Enum):
```
You need to add the url of the source of the color code for Gruvbox for reference. https://github.com/zulip/zulip-terminal/blob/56fd5b27f85f648831ef27c4ddc39c24538d533a/zulipterminal/config/themes.py#L80

Rohitth007 replied:
Oh yes, I'll do that :+1:

rht commented on zulipterminal/config/gruvbox.py:
```diff
@@ -0,0 +1,73 @@
+
+# fmt: off
+class Color(Enum):
+	 BLACK         = 'h234   #1d2021'
```
Though the current version makes it easier for theme contributor to write their colors. Having to specify something like ZTCOLOR('h234', '#1d2021') is a bit more verbose.

neiljp replied:
For simple tuples or even a namedtuple we wouldn't need that prefix.

I agree we need the parentheses and quotes with a tuple approach, but that doesn't seem overly demanding?

Rohitth007 replied:
I think using a string is the cleanest looking of all options. Using a tuple approach would make it use many quotes now that even 16 color name would be included.
Also, is the suggestion to replace enums with tuples/namedtuples? If so it would not be possible to import them for use in other themes. It has to either be a dict or a class. Using a dict would make the STYLE entries less cleaner than the current method.

neiljp replied:
For a tuple it would be like
```python3
class Color(Enum):
    BLACK = ("black", "h234", "#1d2021")
```
If a namedtuple was involved, we'd be able to have `Color.BLACK.in_16_color` or similar.

Rohitth007 replied:
| Tuple | String |
| :-----: | :----: |
| <img src='https://user-images.githubusercontent.com/64144419/121656121-c7390f80-cabc-11eb-87d6-456ec115a7ba.png' height=200> | <img src='https://user-images.githubusercontent.com/64144419/121656213-e59f0b00-cabc-11eb-9d09-8d901302beba.png' height=200> |

Both are fine, depends on ones liking IG.
(I'm leaning towards single string maybe because less quotes reduces clutter?)

rht replied:
> If a namedtuple was involved, we'd be able to have Color.BLACK.in_16_color or similar.

With string, it can always be parsed into a namedtuple by the factory, no problem. The theme config is the user-facing part, where they might want to modify. While it is just small detail, the less cluttered single string definitely reduces the effort for theme contributors, makes it less intimidating.

I do think the tuple version is equally fine. It's just that namedtuple is my least preferred option for the user-facing file.

neiljp replied:
Technically the tuple and string alternatives are not being compared like with like, since urwid colors can have spaces - they don't have underscores except where you've added them here.

I agree the text is less cluttered, and we can always process it. I believe we can eventually make this processing occur in the enum/tuple though.

That said, I'm somewhat wary of going too far into catering for customization since I'd like to just get this iterating and tested as being equivalent to current themes. Once themes are reformatted as multi-depth colors with simple styles and a urwid-palette factory then we can iterate on the representation with less concern.

Note that we can use namedtuple without it needing to be explicit in each enum entry.

preetmishra replied:
Yeah, tuples are more maintainable. It is relatively hard to maintain the spacing and align the strings.

Rohitth007 replied:
The spacing issue is a one time thing right? After that any new change would require only handling one line. Also, the alignment issue is same for tuples also right? Unaligned-tuples wouldn't be very readable.
I'm fine with tuples also though :)

Rohitth007 replied:
I'll mark this as resolved? If Tuples feel more reasonable at a later point we could shift. Also to be considered is using Tuples would mean we have to suppress another flake8 warnings related to "too much spacing in a tuple" I think.

Rohitth007 commented on zulipterminal/cli/run.py:
```diff
@@ -486,10 +487,11 @@ def main(options: Optional[List[str]] = None) -> None:
-            theme_data = theme_with_monochrome_added(THEMES[theme_to_use[0]])
-        else:
-            theme_data = THEMES[theme_to_use[0]]
+        # if color_depth == 1:
+        #     theme_data = theme_with_monochrome_added(THEMES[theme_to_use[0]])
+        # else:
+        #     theme_data = THEMES[theme_to_use[0]]
+        theme_data = generate_theme(theme_to_use[0], color_depth)
```
Yes I'll make that change and cleanup the commit structure. I was waiting for a confirmation of the format to use.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -540,4 +541,70 @@ def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+
+
+if __name__ == "__main__":
+    generate_theme("gruvbox_dark", 1)
```
fixing `msg__bold` to `msg_bold` seems to have fixed this

neiljp replied:
I assume it was some kind of automatic processing issue?

Rohitth007 replied:
Yes I would think so.

Rohitth007 replied:
Yes this happened because I used `pass` instead of `continue` for mono theme. I've fixed it.
The try statement is also removed after the REQUIRED_STYLES update commit.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -540,4 +541,70 @@ def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+if __name__ == "__main__":
+    generate_theme("gruvbox_dark", 1)
+    print()
+    generate_theme("gruvbox_dark", 16)
+    print()
+    generate_theme("gruvbox_dark", 256)
+    print()
+    generate_theme("gruvbox_dark", 2 ** 24)
```
I couldn't observe this for 24bit

neiljp replied:
('msg__bold', '', '', '', '#a89984 , bold', '#1d2021')

Too many underscores :)

Rohitth007 replied:
![image](https://user-images.githubusercontent.com/64144419/121766671-599aeb00-cb71-11eb-8f6b-d1bb25297b10.png)
This is what I see

Rohitth007 replied:
@neiljp Maybe I'll resolve this as I don't think this is a bug? Do tell me if you see this again.

neiljp commented on zulipterminal/config/themes.py:
```diff
@@ -110,6 +112,8 @@
 GRAY24 = "#928374"  # gray_244
 LIGHTMAGENTA24 = "#b16286"  # neutral_purple
 LIGHTMAGENTABOLD24 = f"{LIGHTMAGENTA24}, bold"
+DARKCYAN24 = "#458588"
+BROWN24 = "#b57614"
```
There was no prior comment from where to take these, so perhaps @rht could comment on if these meet expectations?
eg. should darkcyan be aqua rather than neutral_blue? brown is faded_yellow right now.
In any case we should have comments for what these extra colors map to (or come from)

Since we lose these comments in the transfer to the new approach, perhaps it would be useful to rename these colors to match the original names in advance? Hmm, though the conversion script may lose the 16-color interpretation?

Rohitth007 replied:
Yes I was thinking the same thing about using gruvbox names. We do have a 16 color column as well. It has to be manually done.

Re colors, felt these were the closest. I even checked what the color code was called online and it said dark cyan. So, I went with it.

rht replied:
What is the source of this 2 new color codes?

rht replied:
Is it in the gruvbox vim?

Rohitth007 replied:
Yes, from the same link you have provided.

Rohitth007 commented:
Most of the work is done, I just have to add few more tests.
Apart from that, I'm having problems with type hints for the function I wrote in `helper.py`. Enum doesn't seem to work properly with mypy?
Also there were some `flake8` issues caused due to custom indentation and spacing in `Color` and `STYLES`. How do I disable it for those files? It would be nice if even `black` ignores those files completely.

Rohitth007 commented:
All tests are updated, only linting issues are left as mentioned in the above comment.
flake8 issues:
```
W191 indentation contains tabs
E221 multiple spaces before operator
E101 indentation contains mixed spaces and tabs
E122 continuation line missing indentation or outdented
```

Rohitth007 commented:
I have added E221 to the ignore list of flake8

Rohitth007 commented:
I figured it out!! But I had to ignore 1 error as that is a Mypy issue.
```
error: Enum() with dict literal requires string literals
```

Rohitth007 commented:
There are some comments to be resolved but this is ready to be merged I think.

neiljp commented on tests/config/test_themes.py:
```diff
@@ -16,63 +16,13 @@
-    styles_in_theme = {style[0] for style in theme}
-
-    assert len(styles_in_theme) >= len(REQUIRED_STYLES)
-    assert all(required_style in styles_in_theme for required_style in REQUIRED_STYLES)
```
This should be migrated in one commit, not removed in one commit and the new one added in another.

neiljp commented on zulipterminal/themes/_template.py:
```diff
@@ -104,11 +104,14 @@ class DefaultColor(Enum):
     'edit_author'     : 'bold',
     'edit_time'       : 'bold',
+    'current_user'    : '',
+    'muted'           : 'bold',
```
It would be better to improve the test for this and add these earlier rather than later. That test change would be a bugfix of its own.

Then the migration of the test to the new form would be a separate change.

neiljp commented on zulipterminal/config/themes.py:
```diff
@@ -52,11 +52,7 @@ def generate_theme(theme_name: str, color_depth: int) -> ThemeSpec:
-            try:
-                new_style = (style_name, "", "", REQUIRED_STYLES[style_name])
-            except KeyError:
-                continue
```
If the test migration was earlier this change might not be necessary.

That said, we will want to add more error handling to this function later, for situations where the code is updated without running the tests (ie. user modification/exploration of themes) - so you could just leave this in I think?

Rohitth007 replied:
Once we check for equality, this case is covered and shows a incomplete warning. Maybe we can change the wording and explain what incomplete means -
`Warning: Incomplete theme! (either STYLES or REQUIRED_STYLES has missing styles)`?

Rohitth007 replied:
I'm removing try-except for now. Maybe later when we need other error handling we can add them?

Rohitth007 replied:
Do I change the wording for the mentioned Warning?

neiljp reviewed (commented):
@Rohitth007 This is looking really good and the intermediate tests will be useful (though I've not tested them manually), but this is quite a large refactor - bigger than you originally thought, perhaps? :)

The challenge with a larger PR is that lots of smaller tidying/refactoring can be identified along the way, which is what makes this more tricky to handle and get merged as quickly. This will definitely improve our theme handling, we just want to do it well :+1:

Rohitth007 commented on zulipterminal/scripts/theme_reorg.py:
```diff
@@ -1,38 +0,0 @@
-from typing import Any
```
No I had done so, so that after I drop it it'll look like it wasn't even there.
(And also it was the easiest way to make linting pass :))

Rohitth007 commented on zulipterminal/themes/_template.py:
```diff
@@ -0,0 +1,125 @@
+        **{c.name + "__BOLD": c.value + " , bold" for c in DefaultColor},
+    },
+)
+# To add multiple properties use something like:
+#     add_color_prop(EnumClassName, 'BOLD', 'ITALICS', 'STRIKETHROUGH')
```
This was because the helper commit comes after this commit. That's where I change `Enum` -> `add_color_prop`

Rohitth007 replied:
I've moved the helper commit behind so this should not be an issue

Rohitth007 commented on zulipterminal/themes/gruvbox.py:
```diff
@@ -22,13 +24,7 @@ class GruvBoxColor(Enum):
-        **{c.name + "__BOLD": c.value + " , bold" for c in GruvBoxColor},
-    },
-)
+Color = add_color_prop(GruvBoxColor, 'BOLD')
```
I first generate the files in the first few commits and then write the code to use the in later commits. I can't have `helper.py` first because I need a theme to test it with.
I can remove the reference though.

Rohitth007 replied:
How about `color_properties(GruvBoxColor, 'BOLD')` ?

Rohitth007 replied:
Re cyclic import, the issue is themes.py imports all the theme files. If `color_properties` were in themes.py then:
`themes.py -> template/gruvbox -> themes.py`
Error

But now that is in helper.py:
`themes.py -> template/gruvbox -> helper.py`
No problem, though type hinting caused another cyclic import which was solved using TYPE_CHECKING.
If it were in themes.py I would have 2 cyclic import problems.

Rohitth007 replied:
Actually I think I can move the helper behind.

neiljp replied:
Maybe `color_with_extra_properties`?

Rohitth007 replied:
extra would mean there were already some properties?
`color_with_added_properties`?

Rohitth007 replied:
Is the current name `color_properties` fine or is any change needed?

Another thing is that in the current implementation of color_properties doing something like
color_properties(Color, BOLD, ITALICS) would result in only Color.WHITE and Color.WHITE__BOLD_ITALICS but
not Color.WHITE__BOLD or Color.WHITE__ITALICS. One would also have to do
color_properties(Color, BOLD) and color_properties(Color, ITALICS) for the others to work.

I think the current implementation is good for now as other properties are not used and this doesn't restrict its use aswell. Is a FIXME or NOTE enough?

Rohitth007 commented on zulipterminal/scripts/theme_reorg.py:
```diff
@@ -0,0 +1,38 @@
+
+from zulipterminal.config.themes import DEF_base
+
+
```
Yes

Rohitth007 commented on zulipterminal/themes/zt_dark.py:
```diff
@@ -0,0 +1,60 @@
+# ZT DARK __________________________________________________________
+# __________________________________________________________________
```
Oh yes, initially the DefaultColors were in zt_dark, I then moved them to template.

Rohitth007 commented on zulipterminal/themes/zt_blue.py:
```diff
@@ -0,0 +1,60 @@
+# ZT BLUE __________________________________________________________
+# __________________________________________________________________
```
I just thought it looked nice, hence I added them :)
I had seen it in some other conf file or something but we can use docstring too.

Rohitth007 replied:
I've used docstrings now.

Rohitth007 commented on zulipterminal/cli/run.py:
```diff
@@ -486,10 +485,7 @@ def main(options: Optional[List[str]] = None) -> None:
-            theme_data = theme_with_monochrome_added(THEMES[theme_to_use[0]])
-        else:
-            theme_data = THEMES[theme_to_use[0]]
+        theme_data = generate_theme(theme_to_use[0], color_depth)
```
Yes makes sense, so I'll split the `run.py` part of the code and move it to after the test.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -520,4 +520,5 @@ def generate_theme(theme_name: str, color_depth: int) -> ThemeSpec:
             bg = " ".join(bg_codes[2:])
             new_style = (style_name, "", "", "", fg, bg)
         urwid_theme.append(new_style)
+        print(new_style)
```
@neiljp I didn't get your point, this commit would be dropped right?

neiljp replied:
We *could* just drop the test commit, but then we'll have no evidence in the log that the old and new theme results were compared.

Ensuring the tests pass at this commit should guarantee that the new and old way of theme use/generation is the same. If we have a problem later we can look back and check that we didn't lose anything here.

Rohitth007 replied:
Hmm... okay yeah makes sense. :+1:

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -462,6 +463,13 @@
 }
 # fmt: on
 
+ZT_THEMES = {
```
I wanted this to be the new THEMES, hence used a placeholder untill deleting old themes.

Rohitth007 replied:
I've used NEW_THEMES here and later changed it to THEMES

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -6,68 +6,12 @@
-LIGHTMAGENTABOLD24 = f"{LIGHTMAGENTA24}, bold"
-DARKCYAN24 = "#458588"  # neutral_blue
-BROWN24 = "#b57614"  # faded_yellow
+THEMES = {
```
If I name it as THEMES in the previous commits then I would have to move the dict to below the functions.
Then while deleting old THEMES I move it back to the top. Is that fine?

Rohitth007 replied:
Using THEMES causes problems with other tests, so I think NEW_THEMES placeholder is a better option.

Rohitth007 replied:
I'll close this as there is a related discussion for this above.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -484,20 +492,32 @@ def complete_and_incomplete_themes() -> Tuple[List[str], List[str]]:
+        fg, bg = style_conf
+        fg_codes = fg.value.split()
+        bg_codes = bg.value.split()
+        new_style: Tuple[Optional[str], ...] = tuple()
```
I had used the same one used in ThemeSpec though I think a more specific type hint ac be added :+1:

Rohitth007 replied:
StyleSpec added

Rohitth007 commented on zulipterminal/themes/zt_dark.py:
```diff
@@ -0,0 +1,60 @@
+
+
+# fmt: off
+# Use the dot notation to specify a color. Eg: Color.YELLOW
+# Everything after a Dunder(double underscore) are considered
+# as properties. Eg: Color.YELLOW__BOLD
```
Ya I think template has enough information. Wanted to make sure that after the dunder the properties are underscore separated is understood.

Rohitth007 commented on zulipterminal/themes/zt_blue.py:
```diff
@@ -0,0 +1,60 @@
+
+
+# fmt: off
+# Use the dot notation to specify a color. Eg: Color.YELLOW
+# Everything after a Dunder(double underscore) are considered
+# as properties. Eg: Color.YELLOW__BOLD
```
:+1: I'll remove it. Maybe we can direct people to template for further info.

Rohitth007 commented on zulipterminal/themes/_template.py:
```diff
@@ -0,0 +1,125 @@
+# ZT THEME TEMPLATE ________________________________________________
+# __________________________________________________________________
+# Use this template to create a new theme for Zulip Terminal.
```
I was thinking an external document talks about basics of how to make a theme and direct them to the template to understand to specifics of the format etc.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -1,58 +1,9 @@
 from typing import Dict, List, Optional, Tuple
 
+from zulipterminal.themes._template import REQUIRED_STYLES
```
I wanted to make template look like the other themefiles so that we can direct people to template to understand the format.
Having default colors in template make more sense as it tells how a theme like gruvbox should be made, else it would look like you would have to import color from a color file.

neiljp replied:
Having the example theme files should give an idea of how to do it, and gruvbox is an example of separate colors. Having a FAQ entry or a separate short theme writing doc would work fine to describe this.

I'm not saying a template file is a bad idea, just that default colors/styles belong somewhere else.

If someone just edits the template file they'll change default colors/styles and cause problems.

Rohitth007 replied:
It if fine if they edit default color right? Maybe they want to change 1 or 2 colors. Even Gruvbox colors can be edited the same way.

neiljp replied:
My expectation would be that editing a 'template' theme file would only affect the new theme that I'm writing based on the template.

That's not to say that changing some base colors should be difficult, just that this feels like it would be confusing.

Rohitth007 replied:
Hmm...would a name change to `defaults`, clear the confusion? I could move all the doc-strings to a MD file. `defaults` would contain default colors and default mono theme ( technically its the default mono theme for all themes and not Required Styles ).

Rohitth007 replied:
Re script, a simple copy paste of an existing theme would be enough? A script here is too much I think.

Rohitth007 replied:
Having Color in `themes.py` would cause a circular import problem.
themes,py -> zt_dark.py -> themes.py

Rohitth007 replied:
I have made a Common's file in themes where I have placed the default color scheme and required styles.
I have also separated the documentation into a THEME_CONTRIBUTING.md file, not sure where it should go so its currently in the themes folder.
I'm thinking lets put it in the same directory level as README

Rohitth007 commented on zulipterminal/themes/zt_dark.py:
```diff
@@ -0,0 +1,60 @@
+# ZT DARK __________________________________________________________
+# __________________________________________________________________
+# Using Default Color scheme
+from zulipterminal.themes._template import Color
```
It would be nice if all colors were defined i the same way.
But in this method, `DefaultColor -> ColorWithBold -> from file import ColorWithBold as Color`
But if another file wants to use gruvbox colors, `GruvBoxColor -> Color -> from file import Color`
If everything was similar to template everything is easily understood.
Hence also REQUIRED_STYLES makes sense. `_template` would be main file to understand how themefiles should look.

neiljp replied:
Right, but a theme file doesn't have a `REQUIRED_STYLES` dict in it, so having a "template" which has it in feels confusing.

Rohitth007 replied:
I'll close this and we could discuss this above.

Rohitth007 commented:
@neiljp This is ready for another review. I've rebased and made suggested changes.

neiljp reviewed (commented):
@Rohitth007 I've cherry-picked 3 independent commits to reduce the load (with some minor text changes to simplify the diffs and commits), and I hope we're close with the rest :)

Rohitth007 commented on tests/config/test_themes.py:
```diff
@@ -1,10 +1,12 @@
+from enum import Enum
+
```
I'm not sure why that happened. I did see it while working but I thought it was only showing it locally but it also seems to have pushed it. Should git ignore it as `.coverage` exists in `gitignore`?

Rohitth007 replied:
I have removed this.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -516,11 +522,11 @@ def complete_and_incomplete_themes() -> Tuple[List[str], List[str]]:
 
 
-def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+def theme_with_monochrome_added(theme: ThemeSpec) -> Any:
```
I did try this, you meant something like I did in the other function right?
`new_style: StyleSpec = ("", "", "")`
But the slicing step gives Tuple[Any, ...] or Tuple[Union[Any, str, None], ...] as the type.

Rohitth007 commented on tests/config/test_themes.py:
```diff
@@ -36,41 +35,14 @@ def test_all_themes():
 def test_builtin_theme_completeness(theme_name):
     theme = THEMES[theme_name]
-    styles_in_theme = {style[0] for style in theme}
+    theme_styles = theme.STYLES
+    theme_colors = theme.Color
```
If I move this before generating all the tests then I would need to add and update the NEW_THEMES dict at each commit, it would be cleaner to create it only once. I could move it to the commit where new themes were compared with old ones, i.e., the previous commit.

neiljp replied:
I was meaning that you define the new themes, then add the test for completeness, then generate + test, then compare new vs old with the test.

In any case I'd suggest splitting the new theme completeness and removal tests into separate commits.

Rohitth007 commented:
Minor addition: I have added tests to check if all 3 color code are given. See `test_builtin_theme_completeness`

neiljp commented on zulipterminal/config/themes.py:
```diff
@@ -531,3 +537,38 @@ def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+
+    urwid_theme = []
+    for style_name, style_conf in theme_styles.items():
+        fg, bg = style_conf
+        fg_codes = fg.value.split()
+        bg_codes = bg.value.split()
```
You can unpack fg and bg in the `for` statement. It may be clearer not reusing the names later too.

For the fg/bg_codes it does seem like we could unpack them into names instead to improve how the code below reads, ie. avoid the numerical indexing.

This method works fine and we only use it on startup, but note that we split the colors up repeatedly for each style step, so we could improve this later by processing the colors first.

Rohitth007 replied:
Re 1, :+1: 

Re 2, I'm not exactly sure are the number of elements needed to unpack vary if properties are used or not, but I'll check this out.

Re processing colors first, I don't think this takes much time because when the print statement was used. It almost instantly displayed it while loading but the usual things took time, there may be other things to optimize at startup.

neiljp commented on zulipterminal/config/themes.py:
```diff
@@ -531,3 +537,38 @@ def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+
+
+def generate_theme(theme_name: str, color_depth: int) -> ThemeSpec:
+    theme_styles = NEW_THEMES[theme_name].STYLES
```
Let's leave this function to focusing on generating a theme from styles, not the theme name - this makes it independent of the way themes are stored (more decoupled) and easier to test - you don't need the complicated patching in the test.

Rohitth007 replied:
I've moved the rest of the function into `parse_themefile()` which is called in `generate_theme()`.
We could also do other color/theme related stuff like the `add_codehilite_pallette` function into `generate_theme` later.

neiljp replied:
This seems fine for now, though I'm not entirely convinced that we shouldn't just have this inline later.

neiljp commented on tests/config/test_themes.py:
```diff
@@ -49,27 +51,53 @@ def test_complete_and_incomplete_themes():
     assert result == complete_and_incomplete_themes()
 
 
+class theme:
+    class Color(Enum):
+        WHITE__BOLD = "white          #fff   #ffffff , bold"
+        WHITE__BOLD_ITALICS = "white  #fff   #ffffff , bold , italics"
+        DARK_MAGENTA = "dark_magenta  h90    #870087"
+
+    STYLES = {
+        "s1": (Color.WHITE__BOLD, Color.DARK_MAGENTA),
+        "s2": (Color.WHITE__BOLD_ITALICS, Color.DARK_MAGENTA),
+    }
```
Is there a reason the elements of this shouldn't be in the test below?

Note that with the style decoupling you won't need the wrapping class.

Rohitth007 replied:
@neiljp Not sure I get what you mean here about elements, I've decoupled the function though.

neiljp commented on tests/config/test_themes.py:
```diff
@@ -42,6 +71,37 @@ def test_builtin_theme_completeness(theme_name):
+@pytest.mark.parametrize(
+    "theme_name, depth",
+    [
+        ("zt_dark", 16),
```
You've not tested the monochrome bit depth? While that data hasn't been updated, we're migrating from the old function to using generate_theme, so it would be good to include here.

Rohitth007 replied:
Shouldn't the REQUIRED_STYLES completeness tests be enough because this would just compare REQUIRED_STYLES with REQUIRED_STYLES? It wouldn't actually be testing anything right? It would never fail as well - I think?

neiljp replied:
This isn't critical, but seemed missing from the set. I understand it seems "obvious", but many automated tests could be considered that way - though this one more than others perhaps.

Rohitth007 replied:
I've added this check as well

neiljp reviewed (commented):
@Rohitth007 This is really coming together! :) :tada:

We're definitely close to the finishing line - the structure is more stable, and note that my comments are more specific. Also note that the documentation is great to add, and the (improved) theme completeness separately, but just like adding other commits or new parts, it may cause more comments :)

neiljp commented:
> Is the current name `color_properties` fine or is any change needed?

I'm not keen on this name, but I'd like to get this in soon, and we can always iterate from there.

Maybe `extended_colors_with_properties`?

> Another thing is that in the current implementation of color_properties doing something like
> color_properties(Color, BOLD, ITALICS) would result in only Color.WHITE and Color.WHITE__BOLD_ITALICS but
> not Color.WHITE__BOLD or Color.WHITE__ITALICS. One would also have to do
> color_properties(Color, BOLD) and color_properties(Color, ITALICS) for the others to work.
> 
> I think the current implementation is good for now as other properties are not used and this doesn't restrict its use aswell. Is a FIXME or NOTE enough?

I noticed this when reviewing today. Definitely document it, and maybe add a NOTE that this could be extended.

Rohitth007 commented on zulipterminal/config/themes.py:
```diff
@@ -1,64 +1,9 @@
 from typing import Dict, List, Optional, Tuple
 
+from zulipterminal.themes.Common import REQUIRED_STYLES
```
Interesting, having the helper function in color.py would be cleaner I think.

I've used a capital letter because sorting in the directory tree places it to the top.

Is a rename to REQUIRED_STYLE also not required? as it is like a constant definition? and also compares with STYLE

Rohitth007 commented on zulipterminal/themes/Common.py:
```diff
@@ -0,0 +1,93 @@
+    YELLOW        = 'yellow          #ff0       default'
+    LIGHT_BLUE    = 'light_blue      #28d       default'
+    LIGHT_MAGENTA = 'light_magenta   #c8f       default'
+    LIGHT_CYAN    = 'light_cyan      h152       #afd7d7'
```
Since zt_blue used the 16 bit `light_cyan` and `dark_magenta`, I went searching for 256 color codes [here](https://jonasjacek.github.io/colors/). It also had 24bit codes so I used it.

Re "default", It just uses the default foreground or background, so it looks mono with stream colors. We can add proper colors later. It's not related to this PR I think.

The tests don't check if the color codes are correct just that there are 3 space separated words. We should use a **regex** match to check the format maybe.

Rohitth007 replied:
I've added regex tests as well

Rohitth007 commented on zulipterminal/themes/THEME_CONTRIBUTING.md:
````diff
@@ -0,0 +1,89 @@
+```
+
+The actual styles are defined in a dictionary called STYLES
+white takes the style as the key and a tuple with foreground
````
I have no idea why I wrote that. I'll fix that.

Rohitth007 commented on zulipterminal/helper.py:
```diff
@@ -30,6 +32,11 @@
 
 
+if TYPE_CHECKING:  # To avoid circular imports
+    from zulipterminal.themes._template import DefaultColor as Color
```
Wow, how did mypy pass for this. Anyway I'll move this to color.py

neiljp commented on tests/config/test_themes.py:
```diff
@@ -19,12 +21,75 @@
+        assert len(codes) == 3
+        # Check if 16-color alias is correct
+        assert codes[0].replace("_", " ") in aliases_16_color
+        # Check if 24-bit color is #000000-#ffffff
+        assert re.match("(?:#[\\da-f]{6}|default)$", codes[2])
+        # Check if 256 color is #000-#fff or h0-h255 or g0-g100 0r g#00-g#ff
+        code_256 = re.match(
+            "#[\\da-f]{3}|(?:h|g)([\\d]{1,3})|g#[\\da-f]{2}|default$", codes[1]
+        )
+        assert code_256
+        if code_256.group(1) and code_256.group(0).startswith("h"):
+            assert int(code_256.group(1)) < 256
+        elif code_256.group(1) and code_256.group(0).startswith("g"):
+            assert int(code_256.group(1)) <= 100
```
I appreciate the idea here, but I'm not sure if the 256 and 24-bit colors are separate - ie. can the 256 color names be used in 24-bit depth?
@rht Do you know anything about this?

Rohitth007 replied:
Yeah, technically you could use 256 colors and operate them in 24bit mode or use 24bit colors and operate them in 256 color mode where urwid takes care of the conversion internally, but whats the use? We want to use 24bit colors in 24 depth and 256 colors in 256 depth and also down conversion from 24 to 256 is approximate and not good.
I think its easier to search for proper colors online than to rely on conversion.

neiljp replied:
#fff == #ffffff, so some are definitely valid across color depths, and do not lose information. I think we'll want to try and be forgiving with user theme color specification.

That said, these are tests for built-in themes only, so I suppose that's fine for now.

neiljp commented on zulipterminal/config/themes.py:
```diff
@@ -597,3 +602,40 @@ def theme_with_monochrome_added(theme: ThemeSpec) -> ThemeSpec:
+) -> ThemeSpec:
+    urwid_theme = []
+    for style_name, (fg, bg) in theme_styles.items():
+        fg_code16, fg_code256, fg_code24, *fg_props = fg.value.split()
+        bg_code16, bg_code256, bg_code24, *bg_props = bg.value.split()
```
This function reads more clearly now, right?

Rohitth007 replied:
Yes, this is much better now.
We could also reduce `color_depth` to just `depth` to make to more cleaner?

neiljp reviewed (commented):
@Rohitth007 Just some quick feedback as I wasn't sure if it was ready and I'm off to bed.

I couldn't answer all your questions above after I started a review, but I think I answered them below?

Rohitth007 commented on zulipterminal/themes/color.py:
```diff
@@ -0,0 +1,34 @@
+    """
+    prop_n = "_".join([p.upper() for p in prop])
+    prop_v = " , ".join([p.lower() for p in prop])
+    updated_colors: Type[Enum]
+    updated_colors = Enum(  # type: ignore
```
I think these 2 are related to ignoring type hinting:
* https://github.com/python/mypy/issues/529
* https://github.com/python/mypy/issues/5317
I'll add them to the doc string as NOTE?

neiljp replied:
The server repo has a style something like `foo = bar  # type: ignore # explanation` where the explanation can be a link if there's an open report for it. This might be too dynamic to type-check, essentially?

Generally enums don't seem to be type-checked very well using mypy :/
(eg. as of python/typing#535 `.value` is an `Any`)

Rohitth007 replied:
Re too dynamic, yes, I think whatever is inside the Enum functional call is too complicated for mypy to understand.

Rohitth007 commented:
@neiljp I've updated this with the suggestions received.

I've also had to change `Type[Enum]` to Any when I moved colors.py from themes to config. The only other was I can seem to make it work is by importing `Enum` in `__init__.py` of themes and then using it. The problem is I would have to import all of the themefiles too because mypy would be angry and also flake8 is angry because I imported them but didn't use them in `__init__.py`. So I used Any instead.

neiljp commented:
@Rohitth007 I moved that chunk of code between commits we discussed as well as a test that you removed in an earlier commit rather than in the last, and some very minor documentation updates to take into account where color.py is now. After adjusting some commit text I just pushed here and will merge shortly - this will be great to have in, thanks for working on it! :tada:

There are quite a few follow-ups we could explore, though most are not super-important right now and we might file as issues (except the last few):
* default color 256/24-bit improvements (as discussed)
* transparency configuration (see issue/PR)
* new themes? (see other PRs for ideas, but eg. solarized dark/light, gruvbox light?)
* adding FAQ details for adding themes (recommended to discuss with the project rather than add custom ones individually)
* adjust gruvbox colors to reflect the more official names in the palette? (@rht) Maybe also change gruvbox.py to gruvbox_dark.py?
